_source/_posts/2023-09-14-spring-boot-keycloak-kubernetes.md

```diff
@@ -0,0 +1,731 @@
+
+Create a Kubernetes cluster on Google Cloud:
+
+```shell
+gcloud container clusters create jhipster-cluster \
+  --zone southamerica-east1-a \
+  --machine-type n1-standard-4 \
+  --enable-autorepair \
+  --enable-autoupgrade
+```
+
+**Note**: you can choose a different zone and machine type.
+
+If prompted, install the [gke-gcloud-auth-plugin](https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke):
+
+```shell
+gcloud components install gke-gcloud-auth-plugin
+```
+
+Then, fetch the cluster credentials with:
+
+```shell
+gcloud container clusters get-credentials jhipster-cluster --zone southamerica-east1-a
+```
+
+`get-credentials` will update a `kubeconfig` file with appropriate credentials and endpoint information to point `kubectl` at a specific cluster in Google Kubernetes Engine.
+
+Install [cert-manager](https://cert-manager.io/docs/tutorials/getting-started-with-cert-manager-on-google-kubernetes-engine-using-lets-encrypt-for-ingress-ssl/) in your cluster:
+
+```shell
+kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.13.1/cert-manager.yaml
+```
+
+**Note**: If you see a `kubectl` error _The gcp auth plugin has been removed._, see [Here's what to know about changes to kubectl authentication coming in GKE v1.26](https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke) for details on how to fix it.
+
+Apply the deployment descriptors, from the `kubernetes` folder:
+
+```shell
+./kubectl-apply.sh -f
+```
+
+**Important note**: Keycloak client applications (a monolith or microservices) will fail to start until Let's Encrypt has issued the certificate and it has synchronized inside the cluster. In the following section, the issuance process is explained in more detail.
+
+The certificate might take some minutes to be ready, you can check the status by inspecting the object events:
+
+```shell
+kubectl describe certificate keycloak-ssl -n demo
+```
+
+The output should be similar to the following text:
+
+```text
+Events:
+  Type    Reason     Age   From                                       Message
+  ----    ------     ----  ----                                       -------
+  Normal  Issuing    10m   cert-manager-certificates-trigger          Issuing certificate as Secret does not exist
+  Normal  Generated  10m   cert-manager-certificates-key-manager      Stored new private key in temporary Secret resource "keycloak-ssl-rsfkj"
+  Normal  Requested  10m   cert-manager-certificates-request-manager  Created new CertificateRequest resource "keycloak-ssl-mp8s7"
+  Normal  Issuing    10m   cert-manager-certificates-issuing          Issued temporary certificate
+  Normal  Issuing    4m    cert-manager-certificates-issuing          The certificate has been successfully issued
+```
+Also, the ingress service will be updated once the certificate is available. You can check the ingress events with the following command:
+
```

Comment:
I get an error when I try this with the latest `main` branch:

```
  Normal   CreateCertificate  7m8s                  cert-manager-ingress-shim  Successfully created Certificate "keycloak-ssl"
  Warning  Translate          7m4s (x12 over 7m8s)  loadbalancer-controller    Translation failed: invalid ingress spec: could not find service "demo/gateway"
  Normal   Sync               3m54s                 loadbalancer-controller    UrlMap "k8s2-um-ir5cuqmh-demo-gateway-dmiukcj5" created
  Normal   Sync               3m51s                 loadbalancer-controller    TargetProxy "k8s2-tp-ir5cuqmh-demo-gateway-dmiukcj5" created
  Normal   Sync               88s (x4 over 7m8s)    loadbalancer-controller    Scheduled for sync
  Warning  Sync               81s (x5 over 3m50s)   loadbalancer-controller    Error syncing to GCP: error running load balancer syncing routine: loadbalancer ir5cuqmh-demo-gateway-dmiukcj5 does not exist: googleapi: Error 412: Constraint constraints/compute.restrictLoadBalancerCreationForTypes violated for projects/raibleward. Forwarding Rule projects/raibleward/global/forwardingRules/k8s2-fr-ir5cuqmh-demo-gateway-dmiukcj5 of type EXTERNAL_HTTP_HTTPS is not allowed., conditionNotMet
```

I'll delete my cluster and try again.

Reply:
The load balancer creation might be failing because of restrictions on the project I'm using. https://cloud.google.com/load-balancing/docs/org-policy-constraints

I'm going to create a new project and try it there.

Reply:
It worked when I created a new project and cluster!